extractor/filesystem/os/nix/nix.go

```diff
@@ -0,0 +1,284 @@
+	var matches []string
+	// packageStoreRegex implements the regex for the /nix/store pattern.
+	packageStoreRegex := regexp.MustCompile(`^([a-zA-Z0-9]{32})-([a-zA-Z0-9.-]+)-([0-9.]+)(?:-(\S+))?$`)
+	packageStoreUnstableRegex := regexp.MustCompile(`^([a-zA-Z0-9]{32})-([a-zA-Z0-9.-]+)-(unstable-[0-9]{4}-[0-9]{2}-[0-9]{2})$`)
+
+	pkgs := []*extractor.Inventory{}
+
+	m, err := osrelease.GetOSRelease(input.FS)
+	if err != nil {
+		log.Errorf("osrelease.GetOSRelease(): %v", err)
+	}
+
+	// Return if canceled or exceeding deadline.
+	if err := ctx.Err(); err != nil {
+		return pkgs, fmt.Errorf("%s halted at %q because of context error: %v", e.Name(), input.Path, err)
+	}
+
+	// e.g.
+	// pkg = nix/store/mccadsxm3sgrcx2z0b9c8d5wgnlg7a3z-libarchive-3.7.4-lib/lib/libarchive.so.13.7.4
+	// matches = [mccadsxm3sgrcx2z0b9c8d5wgnlg7a3z-libarchive-3.7.4-lib mccadsxm3sgrcx2z0b9c8d5wgnlg7a3z libarchive 3.7.4 lib]
+	pkg := strings.Split(input.Path, "/")[2]
+
+	// Handle unstable packages
+	if !strings.Contains(pkg, "unstable") {
+		matches = packageStoreRegex.FindStringSubmatch(pkg)
+	} else {
+		matches = packageStoreUnstableRegex.FindStringSubmatch(pkg)
+	}
+
+	if len(matches) != 0 {
+		pkgHash := matches[1]
+		pkgName := matches[2]
+		pkgVersion := matches[3]
+
+		if pkgHash != "" && pkgName != "" && pkgVersion != "" {
+			i := &extractor.Inventory{
+				Name:    pkgName,
+				Version: pkgVersion,
+				Metadata: &Metadata{
+					PackageName:       pkgName,
+					PackageVersion:    pkgVersion,
+					PackageHash:       pkgHash,
+					OSID:              m["ID"],
+					OSVersionCodename: m["VERSION_CODENAME"],
+					OSVersionID:       m["VERSION_ID"],
+				},
+				Locations: []string{input.Path},
+			}
+
+			if len(matches) > 4 {
+				pkgOutput := matches[4]
+				i.Metadata.(*Metadata).PackageOutput = pkgOutput
+			}
+
+			pkgs = append(pkgs, i)
+		}
+	}
+
+	return pkgs, nil
+}
+
+// ToPURL converts an inventory created by this extractor into a PURL.
+func (e Extractor) ToPURL(i *extractor.Inventory) *purl.PackageURL {
+	m := i.Metadata.(*Metadata)
+	q := map[string]string{}
+	distro := toDistro(m)
+
+	if distro != "" {
+		q[purl.Distro] = distro
+	}
+
+	return &purl.PackageURL{
+		Type:       purl.TypeNix,
+		Name:       i.Name,
+		Version:    i.Version,
+		Namespace:  toNamespace(m),
+		Qualifiers: purl.QualifiersFromMap(q),
+	}
+}
+
+// Ecosystem returns the OSV Ecosystem of the software extracted by this extractor.
+func (Extractor) Ecosystem(i *extractor.Inventory) string {
+	m := i.Metadata.(*Metadata)
+	osID := cases.Title(language.English).String(toNamespace(m))
+
+	if m.OSVersionID == "" {
+		return osID
+	}
+	return osID + ":" + m.OSVersionID
+}
+
+func toNamespace(m *Metadata) string {
+	if m.OSID != "" {
+		return m.OSID
+	}
+	log.Errorf("os-release[ID] not set, fallback to 'nixos'")
+	return "nixos"
+}
+
+func toDistro(m *Metadata) string {
```

Comment:
Does this work on Nix?

Reply:
As mentioned in another comment, NixOS does provide the OSID, VERSION_ID, and VERSION_CODENAME fields in its os-release file. The `toDistro` function should work as expected

Reply:
sg